kernel/mellanox_ethernet/src/command_queue.rs

```diff
@@ -0,0 +1,740 @@
+        write!(f, "CQE:output_mailbox_pointer_l: {:#X} \n", self.output_mailbox_pointer_l.read().get())?;
+        write!(f, "CQE:output_length: {} \n", self.output_length.read().get())?;
+        write!(f, "CQE:token_signature_status_own: {:#X} \n", self.token_signature_status_own.read().get())
+    }
+}
+
+impl CommandQueueEntry {
+    /// Sets the type of transport which carries the command as PCIe.
+    fn set_type_of_transport(&mut self) {
+        self.type_of_transport.write(U32::new(CommandTransportType::PCIe as u32));
+    }
+
+    /// Sets length of input data in bytes.
+    /// This value is different for every command, and can be taken from Chapter 23 of the PRM. 
+    fn set_input_length_in_bytes(&mut self, length: u32) {
+        self.input_length.write(U32::new(length))
+    }
+
+    /// Sets the first 16 bytes of input data that are written inline in the command.
+    /// The valid values for each field are different for every command, and can be taken from Chapter 23 of the PRM.
+    ///
+    /// # Arguments
+    /// * `opcode`: value identifying which command has to be carried out
+    /// * `opmod`: opcode modifier. If None, field will be set to zero.
+    /// * `command0`: the first 4 bytes of actual command data. If None, field will be set to zero.
+    /// * `command1`: the second 4 bytes of actual command data. If None, field will be set to zero. 
+    fn set_input_inline_data(&mut self, opcode: CommandOpcode, opmod: Option<u16>, command0: Option<u32>, command1: Option<u32>) {
+        self.command_input_opcode.write(U32::new((opcode as u32) << 16));
+        self.command_input_opmod.write(U32::new(opmod.unwrap_or(0) as u32));
+        self.command_input_inline_data_0.write(U32::new(command0.unwrap_or(0)));
+        self.command_input_inline_data_1.write(U32::new(command1.unwrap_or(0)));
+    }
+
+    /// Returns the value written to the input opcode field of the command
+    fn get_command_opcode(&self) -> CommandOpcode {
+        match self.command_input_opcode.read().get() >> 16 {
+            0x100   => {CommandOpcode::QueryHcaCap},
+            0x101   => {CommandOpcode::QueryAdapter}, 
+            0x102   => {CommandOpcode::InitHca}, 
+            0x103   => {CommandOpcode::TeardownHca}, 
+            0x104   => {CommandOpcode::EnableHca}, 
+            0x105   => {CommandOpcode::DisableHca}, 
+            0x107   => {CommandOpcode::QueryPages}, 
+            0x108   => {CommandOpcode::ManagePages}, 
+            0x10A   => {CommandOpcode::QueryIssi}, 
+            0x10B   => {CommandOpcode::SetIssi}, 
+            _       => {CommandOpcode::Unknown}
+        }
+    }
+
+    /// Returns the first 16 bytes of output data that are written inline in the command.
+    /// The valid values for each field are different for every command, and can be taken from Chapter 23 of the PRM.
+    fn get_output_inline_data(&self) -> (u8, u32, u32, u32) {
```

Comment:
could be confusing to deal with 4-element tuples of similar values.... perhaps place these into a struct? Depends on usage, I suppose.

Reply:
The bits that are actually going to be used from each field are different for each command. It's a private function and is only used within other functions of the command queue that then extract the relevant bit fields.
I don't personally think it's too confusing, but it wouldn't take much effort to change this to a struct.

Reply:
Your choice. If the valid values change per command, I'd recommend the use of an enum to represent the possible sets of valid values, such that you can more strongly enforce correctness using typing. The raw values could be converted into that enum and then returned from either this function or a surrounding one.